### PR TITLE
Prerequisite & Verification for puppet environment

### PR DESCRIPTION
**Pull request by dubewarsagar**
Add a Prerequisite & Verification steps.
Prerequisite to create a back of `/etc/puppetlabs/code/environments` before upgrade.
Verification module to verify contents of the `/etc/puppetlabs/code/environments` file, post upgrade.
The same has been requested in BZ-2097843

https://bugzilla.redhat.com/show_bug.cgi?id=2097843

* [x] I am familiar with the [contributing](https://github.com/theforeman/foreman-documentation/blob/master/CONTRIBUTING.md) guidelines.

Please cherry-pick my commits into:

* [ ] Foreman 3.5/Katello 4.7 (planned Satellite 6.13)
* [ ] Foreman 3.4/Katello 4.6 (EL8 only)
* [ ] Foreman 3.3/Katello 4.5 on EL7 & EL8 (Satellite 6.12 on EL8 only)
* [ ] Foreman 3.2/Katello 4.4 on EL7 & EL8
* [ ] Foreman 3.1/Katello 4.3 on EL7 & EL8 (Satellite 6.11 EL7/8, orcharhino 6.1 on EL7, orcharhino 6.2 on EL7/8)
* For Foreman 3.0 or older, please create a separate PR.
* We do not accept PRs for Foreman 2.3 or older.

### CODE REVIEW

**Comment by adamlazik1** [comment]
Wouldn't it be better to add verification under `.Verification` section instead of creating a new file?

**Comment by dubewarsagar** [comment]
> Wouldn't it be better to add verification under `.Verification` section instead of creating a new file?

As per my understanding of the structure of this document, it is more appropriate to keep this as a separate section. There is no one single section, which can accommodate this information, hence it is kept in a separate section.

Regards,
Sagar Dubewar

File: guides/common/modules/proc_validating-puppet-environments-on-the-external-capsule-servers.adoc [hunk 1]
```diff
@@ -0,0 +1,19 @@
+:_content-type: PROCEDURE
+
+[id="validating-puppet-environments-on-the-external-capsule-servers_{context}"]
```
Comment (adamlazik1) [inline guides/common/modules/proc_validating-puppet-environments-on-the-external-capsule-servers.adoc:3]:
```suggestion
[id="validating-puppet-environments-on-the-external-{smart-proxy-context}-servers_{context}"]
```

File: guides/common/modules/proc_validating-puppet-environments-on-the-external-capsule-servers.adoc [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+= Validating Puppet Environments on the External {SmartProxyServer}
+
+After upgrading the {SmartProxyServer} to {ProjectVersion}, ensure the availability of the puppet environments inside the `/etc/puppetlabs/code/environments` file of the {SmartProxyServer}.
```
Comment (adamlazik1) [inline guides/common/modules/proc_validating-puppet-environments-on-the-external-capsule-servers.adoc:6]:
```suggestion
After upgrading the {SmartProxyServer} to {ProjectVersion}, ensure the puppet environments are available on the {SmartProxyServer}.
```

File: guides/common/modules/proc_validating-puppet-environments-on-the-external-capsule-servers.adoc [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+.Procedure
+. To view the Puppet environments, open the `/etc/puppetlabs/code/environments` file.
++
+[options="nowrap", subs="+quotes,verbatim,attributes"]
+----
+# vim /etc/puppetlabs/code/environments
+----
+. If any of the puppet environments are missing, copy them back from:
+.. the {project} server or 
+.. the backup taken prior to upgrade
```
Comment (adamlazik1) [inline guides/common/modules/proc_validating-puppet-environments-on-the-external-capsule-servers.adoc:17]:
```suggestion
. If any of the puppet environments are missing, copy them back from the {ProjectServer} or the backup taken prior to upgrade.
```

File: guides/common/modules/proc_validating-puppet-environments-on-the-external-capsule-servers.adoc [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+. To view the Puppet environments, open the `/etc/puppetlabs/code/environments` file.
++
+[options="nowrap", subs="+quotes,verbatim,attributes"]
+----
+# vim /etc/puppetlabs/code/environments
+----
+. If any of the puppet environments are missing, copy them back from:
+.. the {project} server or 
+.. the backup taken prior to upgrade 
+. Ensure to have the right permissions and ownership of the missing contents inside  `/etc/puppetlabs/code/environments/` file after they have been restored back.
```
Comment (adamlazik1) [inline guides/common/modules/proc_validating-puppet-environments-on-the-external-capsule-servers.adoc:18]:
```suggestion
. Ensure to have the right permissions and ownership of the missing contents inside  the `/etc/puppetlabs/code/environments/` file after they have been restored back.
```

File: guides/doc-Upgrading_and_Updating/master.adoc [hunk 5]
```diff
@@ -88,6 +88,9 @@ include::topics/post_upgrade_template_checks.adoc[leveloffset=+2]
 //Tuning Satellite Server with Predefined Profiles
 include::common/modules/proc_tuning-with-predefined-profiles.adoc[leveloffset=+2]
 
+//Validating Puppet Environments on the External Capsule Servers
+include::common/modules/proc_validating-puppet-environments-on-the-external-capsule-servers.adoc[leveloffset=+2]
```
Comment (adamlazik1) [inline guides/doc-Upgrading_and_Updating/master.adoc:92]:
```suggestion
include::common/modules/proc_validating-puppet-environments-on-the-external-smartproxy-servers.adoc[leveloffset=+2]
```
Please also rename the file.

**Review by adamlazik1 — COMMENTED** [review]
A few suggestions.

**Comment by dubewarsagar** [comment]
> A few suggestions.

Done.

**Comment by adamlazik1** [comment]
Please don't forget to change the file name, otherwise the guide won't build properly.

File: guides/doc-Upgrading_and_Updating/master.adoc [hunk 6]
```diff
@@ -88,6 +88,9 @@ include::topics/post_upgrade_template_checks.adoc[leveloffset=+2]
 //Tuning Satellite Server with Predefined Profiles
 include::common/modules/proc_tuning-with-predefined-profiles.adoc[leveloffset=+2]
 
+//Validating Puppet Environments on the External Capsule Servers
```
Comment (adamlazik1) [inline guides/doc-Upgrading_and_Updating/master.adoc:91]:
```suggestion
//Validating Puppet Environments on the External SmartProxy Servers
```

**Comment by dubewarsagar** [comment]
> Please don't forget to change the file name, otherwise the guide won't build properly.

Done.

**Review by adamlazik1 — APPROVED** [review]
LGTM

**Comment by adamlazik1** [comment]
Merged into 2.5.

**Pull request merged**